streaming/src/main/scala/org/apache/spark/streaming/dstream/TrackStateDStream.scala

```diff
@@ -132,22 +132,39 @@ class InternalTrackStateDStream[K: ClassTag, V: ClassTag, S: ClassTag, E: ClassT
   /** Method that generates a RDD for the given time */
   override def compute(validTime: Time): Option[RDD[TrackStateRDDRecord[K, S, E]]] = {
     // Get the previous state or create a new empty state RDD
-    val prevStateRDD = getOrCompute(validTime - slideDuration).getOrElse {
-      TrackStateRDD.createFromPairRDD[K, V, S, E](
-        spec.getInitialStateRDD().getOrElse(new EmptyRDD[(K, S)](ssc.sparkContext)),
-        partitioner, validTime
-      )
+    val prevStateRDD = getOrCompute(validTime - slideDuration) match {
+      case Some(rdd) =>
+        if (rdd.partitioner != Some(partitioner)) {
+          // If the RDD is not partitioned the right way, let us repartition it using the
+          // partition index as the key. This is to ensure that state RDD is always partitioned
+          // before creating another state RDD using it
+          val kvRDD = rdd.mapPartitions { iter =>
+            iter.map { x => (TaskContext.get().partitionId(), x)}
+          }
+          kvRDD.partitionBy(partitioner).mapPartitions(iter => iter.map { _._2 },
```

Comment:
If the user specifies a custom `partitioner`, we cannot use `partitioner` on the `partitionId`s.